Hazel/Udp/UdpConnection.Fragmented.cs

```diff
@@ -0,0 +1,241 @@
+        {
+            MtuTest(_mtuIndex);
+        }
+
+        private void MtuTest(byte index)
+        {
+            var mtu = PossibleMtu[index];
+            var failed = false;
+
+            var buffer = new byte[mtu];
+            buffer[0] = (byte)UdpSendOption.MtuTest;
+            var id = AttachReliableID(buffer, 1, () =>
+            {
+                if (failed) return;
+                MtuOk(index);
+            });
+            buffer[mtu - 2] = (byte)mtu;
+            buffer[mtu - 1] = (byte)(mtu >> 8);
+
+            WriteBytesToConnection(buffer, buffer.Length, () =>
+            {
+                failed = true;
+                AcknowledgeMessageId(id);
+
+                if (index == 0)
+                {
+                    DisconnectInternal(HazelInternalErrors.MtuTooLow, "Connection MTU is lower than the minimum");
+                }
+            });
+        }
+
+        private void MtuOk(byte index)
+        {
+            _mtuIndex = index;
+            _mtu = PossibleMtu[index];
+            MtuChanged?.Invoke();
+
+            if (_mtuIndex < PossibleMtu.Length - 1)
+            {
+                MtuTest((byte)(index + 1));
+            }
+        }
+
+        private void MtuTestMessageReceive(MessageReader message)
+        {
+            message.Position = message.Length - 2;
+            var mtu = message.ReadUInt16();
+
+            if (mtu != message.Length)
+            {
+                return;
+            }
+
+            ProcessReliableReceive(message.Buffer, 1, out _);
+        }
+
+        private const byte FragmentHeaderSize = sizeof(byte) + sizeof(ushort) + sizeof(ushort) + sizeof(ushort);
+
+        protected void FragmentedSend(byte sendOption, byte[] data, Action ackCallback, bool includeHeader)
+        {
+            var length = includeHeader ? data.Length + 1 : data.Length;
+
+            var id = (ushort)Interlocked.Increment(ref _lastFragmentedId);
+            var fragmentSize = Mtu;
+            var fragmentDataSize = fragmentSize - FragmentHeaderSize;
+            var fragmentsCount = (int)Math.Ceiling(length / (double)fragmentDataSize);
+
+            if (fragmentsCount >= ushort.MaxValue)
+            {
+                throw new HazelException("Too many fragments");
+            }
+
+            var acksReceived = 0;
+
+            for (ushort i = 0; i < fragmentsCount; i++)
+            {
+                var dataLength = Math.Min(fragmentDataSize, length - fragmentDataSize * i);
+                var buffer = new byte[dataLength + FragmentHeaderSize];
+
+                buffer[0] = (byte)UdpSendOption.Fragment;
+
+                AttachReliableID(buffer, 1, () =>
+                {
+                    acksReceived++;
+
+                    if (acksReceived >= fragmentsCount)
+                    {
+                        ackCallback?.Invoke();
+                    }
+                });
+
+                buffer[3] = (byte)fragmentsCount;
```

Comment:
tbh, kinda tempted to say only 1 byte. No one should use Hazel to transfer 128kb (500byte MTU * 256 parts) much less 32mb.